README.md
# p6-cirrus-inc


## Table of Contents


### p6-cirrus-inc
- [p6-cirrus-inc](#p6-cirrus-inc)
  - [Badges](#badges)
  - [Distributions](#distributions)
  - [Summary](#summary)
  - [Contributing](#contributing)
  - [Code of Conduct](#code-of-conduct)
  - [Changes](#changes)
    - [Usage](#usage)
  - [Author](#author)

### Badges

[![License](https://img.shields.io/badge/License-Apache%202.0-yellowgreen.svg)](https://opensource.org/licenses/Apache-2.0)
[![Gitpod ready-to-code](https://img.shields.io/badge/Gitpod-ready--to--code-blue?logo=gitpod)](https://gitpod.io/#https://github.com/p6m7g8/p6-cirrus-inc)
[![Mergify](https://img.shields.io/endpoint.svg?url=https://gh.mergify.io/badges/p6m7g8/p6-cirrus-inc/&style=flat)](https://mergify.io)
[![codecov](https://codecov.io/gh/p6m7g8/p6-cirrus-inc/branch/master/graph/badge.svg?token=14Yj1fZbew)](https://codecov.io/gh/p6m7g8/p6-cirrus-inc)
[![Known Vulnerabilities](https://snyk.io/test/github/p6m7g8/p6-cirrus-inc/badge.svg?targetFile=package.json)](https://snyk.io/test/github/p6m7g8/p6-cirrus-inc?targetFile=package.json)

## Summary

## Contributing

- [How to Contribute](CONTRIBUTING.md)

## Code of Conduct

- [Code of Conduct](CODE_OF_CONDUCT.md)

## Changes

- [Change Log](CHANGELOG.md)

### Usage

#### init.zsh:

- p6df::modules::p6-cirrus-inc::deps()
- p6df::modules::p6-cirrus-inc::init()

#### app.sh:

- p6_cirrius_inc_sg_myself_allow(sg_name, [port=443])
- p6_cirrus_inc_cleanup()
- p6_cirrus_inc_instance_amazon_create()
- p6_cirrus_inc_instance_bastion_create()
- p6_cirrus_inc_instance_create(name, ami_id, [instance_type=t3a.nano], [user_data=], [subnet_type=infra])
- p6_cirrus_inc_instance_irc_create()
- p6_cirrus_inc_instance_jenkins_create()
- p6_cirrus_inc_instance_rhel8_create()
- p6_cirrus_inc_instance_ubuntu18_create()
- p6_cirrus_inc_sg_link_bastion_vpc_ssh()
- p6_cirrus_inc_sg_link_bastion_world_ssh()
- p6_cirrus_inc_sg_link_outbound_world_world()
- str key_id = p6_aws_kms_svc_key_make(account_id, key_description, key_alias)
- str sg_bastion_ssh_id = p6_cirrus_inc_sg_bastion_ssh_create([vpc_id=$AWS_VPC])
- str sg_instance_ssh_id = p6_cirrus_inc_sg_instance_ssh_create([vpc_id=$AWS_VPC])
- str sg_outbound_id = p6_cirrus_inc_sg_outbound_ssh_create([vpc_id=$AWS_VPC])


## Author

Philip M . Gollucci <pgollucci@p6m7g8.com>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.1.0](https://github.com/p6m7g8/p6-cirrus-inc/compare/v2.0.0...v2.1.0) (2020-12-06)


### Features

* **api:** centralize ::version() ([68c05d6](https://github.com/p6m7g8/p6-cirrus-inc/commit/68c05d6607d9398e71d50be3ebab571f46c702e4))
* **community:** adds CODE_OF_CONDUCT ([bcc350a](https://github.com/p6m7g8/p6-cirrus-inc/commit/bcc350a370adec2607bfbbe88ab835add129cf88))
* **github:** adds ISSUE_TEMPLATEs ([826d278](https://github.com/p6m7g8/p6-cirrus-inc/commit/826d2789a13868b2bd3f57fc8483a28a583cef53))
* **mergify:** re-gens ([9d8b292](https://github.com/p6m7g8/p6-cirrus-inc/commit/9d8b2926047e62e66955db7edc28312ce332d957))


* **docs:** fixes whitespace ([4a07c1f](https://github.com/p6m7g8/p6-cirrus-inc/commit/4a07c1f4e388217777e448f5b598c47c756be26f))
* **README:** updates to template v2.0 ([bcc852b](https://github.com/p6m7g8/p6-cirrus-inc/commit/bcc852b0a1c5023d5be1fcc02ced86177a195cb2))
* **semantic:** no longer needed, defaults are fine ([69d13f3](https://github.com/p6m7g8/p6-cirrus-inc/commit/69d13f39e8cda7e1494c185ef75a1e0b7d57f1f5))

## [2.0.0](https://github.com/p6m7g8/p6-cirrus-inc/compare/v1.0.1...v2.0.0) (2020-10-11)

### 1.0.1 (2020-09-20)


### Features

* **sgs:** allow 'my' ip/32 ([#2](https://github.com/p6m7g8/p6-cirrus-inc/issues/2)) ([f0dab65](https://github.com/p6m7g8/p6-cirrus-inc/commit/f0dab6520f0d33f9162ffb2b05860220ca1cb516))


* **actions:** adds semantic.yml and mergiy configs ([1786621](https://github.com/p6m7g8/p6-cirrus-inc/commit/178662166990f71e8734fa4a30d4e0033bbf3c0e))
* **whitespace:** regens docs ([#3](https://github.com/p6m7g8/p6-cirrus-inc/issues/3)) ([f0064d0](https://github.com/p6m7g8/p6-cirrus-inc/commit/f0064d002ab2d0cc2ef84d1d48b73b6bb9e24ba8))
* **zsh:** makes loadable ([#1](https://github.com/p6m7g8/p6-cirrus-inc/issues/1)) ([7382f39](https://github.com/p6m7g8/p6-cirrus-inc/commit/7382f397a51560eac63e29ae642f9e9de59d3444))


lib/app.sh
######################################################################
#<
#
# Function: p6_cirrus_inc_instance_create(name, ami_id, [instance_type=t3a.nano], [user_data=], [subnet_type=infra])
#
#  Args:
#	name -
#	ami_id -
#	OPTIONAL instance_type - [t3a.nano]
#	OPTIONAL user_data - []
#	OPTIONAL subnet_type - [infra]
#
#>
######################################################################
p6_cirrus_inc_instance_create() {
    local name="$1"
    local ami_id="$2"
    local instance_type="${3:-t3a.nano}"
    local user_data="${4:-}"
    local subnet_type="${5:-infra}"

    [ -z "$ami_id" ] && ami_id=$(p6_cirrus_inc_amis_freebsd12_latest)
    [ -n "$user_data" ] && user_data="--user-data=$user_data"

    local key_name=$(p6_cirrus_inc_key_pair_make "$USER")

    local sg_id
    local subnet_id
    case $name in
    bastion)
        sg_id=$(p6_cirrus_inc_sg_id_from_group_name "bastion-ssh")
        subnet_id=$(p6_cirrus_inc_subnet_bastion_get)
        ;;
    *)
        sg_id=$(p6_cirrus_inc_sg_id_from_group_name "vpc-ssh")
        subnet_id=$(p6_cirrus_inc_subnet_${subnet_type}_get)
        ;;
    esac

    local sg_outbound_id=$(p6_cirrus_inc_sg_id_from_group_name "outbound")

    p6_aws_ec2_instances_run \
        --output json \
        --key-name $key_name \
        --image-id $ami_id \
        --instance-type $instance_type \
        --security-group-ids $sg_id $sg_outbound_id \
        --subnet-id $subnet_id \
        $user_data \
        --tag-specifications "'ResourceType=instance,Tags=[{Key=Name,Value=$name}]'"
}

######################################################################
#<
#
# Function: p6_cirrius_inc_sg_myself_allow(sg_name, [port=443])
#
#  Args:
#	sg_name -
#	OPTIONAL port - [443]
#
#>
######################################################################
p6_cirrius_inc_sg_myself_allow() {
    local sg_name="$1"
    local port="${2:-443}"

    local myip=$(dig +short myip.opendns.com @resolver1.opendns.com)

    local sg_id=$(p6_cirrus_inc_sg_id_from_group_name "$sg_name")
    p6_aws_ec2_security_group_ingress_authorize --group-id $sg_id --protocol tcp --port $port --cidr $myip/32
}

######################################################################
#<
#
# Function: p6_cirrus_inc_instance_ubuntu18_create()
#
#>
######################################################################
p6_cirrus_inc_instance_ubuntu18_create() {

    p6_cirrus_inc_instance_create "ubuntu18"
}

######################################################################
#<
#
# Function: p6_cirrus_inc_instance_rhel8_create()
#
#>
######################################################################
p6_cirrus_inc_instance_rhel8_create() {

    p6_cirrus_inc_instance_create "rhel-8"
}

######################################################################
#<
#
# Function: p6_cirrus_inc_instance_amazon_create()
#
#>
######################################################################
p6_cirrus_inc_instance_amazon_create() {

    p6_cirrus_inc_instance_create "amazon2"
}

######################################################################
#<
#
# Function: p6_cirrus_inc_cleanup()
#
#>
######################################################################
p6_cirrus_inc_cleanup() {

    true
    ## terminate instances
    ## delete sgs
}

######################################################################
#<
#
# Function: str sg_bastion_ssh_id = p6_cirrus_inc_sg_bastion_ssh_create([vpc_id=$AWS_VPC])
#
#  Args:
#	OPTIONAL vpc_id - [$AWS_VPC]
#
#  Returns:
#	str - sg_bastion_ssh_id
#
#>
######################################################################
p6_cirrus_inc_sg_bastion_ssh_create() {
    local vpc_id=${1:-$AWS_VPC}

    local sg_bastion_ssh_id=$(p6_aws_ec2_security_group_create "'Allows SSH TPC(22) Inbound From Anywhere'" "bastion-ssh" --vpc-id $vpc_id --output text)

    p6_aws_ec2_tags_create "$sg_bastion_ssh_id" "Key=Name,Value=bastion-ssh"

    p6_return_str "$sg_bastion_ssh_id"
}

######################################################################
#<
#
# Function: str sg_instance_ssh_id = p6_cirrus_inc_sg_instance_ssh_create([vpc_id=$AWS_VPC])
#
#  Args:
#	OPTIONAL vpc_id - [$AWS_VPC]
#
#  Returns:
#	str - sg_instance_ssh_id
#
#>
######################################################################
p6_cirrus_inc_sg_instance_ssh_create() {
    local vpc_id=${1:-$AWS_VPC}

    local sg_instance_ssh_id=$(p6_aws_ec2_security_group_create "'Allows SSH TPC(22) Inbound From Bastion Only'" "vpc-ssh" --vpc-id $vpc_id --output text)

    p6_aws_ec2_tags_create "$sg_instance_ssh_id" "Key=Name,Value=vpc-ssh"

    p6_return_str "$sg_instance_ssh_id"
}

######################################################################
#<
#
# Function: str sg_outbound_id = p6_cirrus_inc_sg_outbound_ssh_create([vpc_id=$AWS_VPC])
#
#  Args:
#	OPTIONAL vpc_id - [$AWS_VPC]
#
#  Returns:
#	str - sg_outbound_id
#
#>
######################################################################
p6_cirrus_inc_sg_outbound_ssh_create() {
    local vpc_id=${1:-$AWS_VPC}

    local sg_outbound_id=$(p6_aws_ec2_security_group_create "'Allows All TCP Outbound to ALL'" "outbound" --vpc-id $vpc_id --output text)

    p6_aws_ec2_tags_create "$sg_outbound" "Key=Name,Value=outbound"

    p6_return_str "$sg_outbound_id"
}

######################################################################
#<
#
# Function: p6_cirrus_inc_sg_link_outbound_world_world()
#
#>
######################################################################
p6_cirrus_inc_sg_link_outbound_world_world() {

    local sg_outbound_id=$(p6_cirrus_inc_sg_id_from_group_name "outbound")

    p6_aws_ec2_security_group_egress_authorize $sg_outbound_id --protocol tcp --port 0 --cidr 0.0.0.0/0
}

######################################################################
#<
#
# Function: p6_cirrus_inc_sg_link_bastion_world_ssh()
#
#>
######################################################################
p6_cirrus_inc_sg_link_bastion_world_ssh() {

    local sg_bastion_ssh_id=$(p6_cirrus_inc_sg_id_from_group_name "bastion-ssh")
    p6_aws_ec2_security_group_ingress_authorize --group-id $sg_bastion_ssh_id --protocol tcp --port 22 --cidr 0.0.0.0/0
}

## Links: a_b_proto
######################################################################
#<
#
# Function: p6_cirrus_inc_sg_link_bastion_vpc_ssh()
#
#>
######################################################################
p6_cirrus_inc_sg_link_bastion_vpc_ssh() {

    local sg_bastion_ssh_id=$(p6_cirrus_inc_sg_id_from_group_name "bastion-ssh")
    local sg_instance_ssh_id=$(p6_cirrus_inc_sg_id_from_group_name "vpc-ssh")

    # bastion->vpc
    p6_aws_ec2_security_group_egress_authorize $sg_instance_ssh_id --protocol tcp --port 22 --source-group $sg_bastion_ssh_id

    # bastion<-vpc
    p6_aws_ec2_security_group_ingress_authorize --group-id $sg_bastion_ssh_id --protocol tcp --port 22 --source-group $sg_instance_ssh_id

    # vpc->bastion
    p6_aws_ec2_security_group_egress_authorize $sg_bastion_ssh_id --protocol tcp --port 22 --source-group $sg_instance_ssh_id
}

######################################################################
#<
#
# Function: p6_cirrus_inc_instance_bastion_create()
#
#>
######################################################################
p6_cirrus_inc_instance_bastion_create() {

    p6_cirrus_inc_instance_create "bastion"
}

######################################################################
#<
#
# Function: p6_cirrus_inc_instance_irc_create()
#
#>
######################################################################
p6_cirrus_inc_instance_irc_create() {

    local ami_id=$(p6_cirrus_inc_amis_freebsd12_latest)
    p6_cirrus_inc_instance_create "irc" "$ami_id" "t2.micro"
}

######################################################################
#<
#
# Function: p6_cirrus_inc_instance_jenkins_create()
#
#>
######################################################################
p6_cirrus_inc_instance_jenkins_create() {

    local ami_id=$(p6_cirrus_inc_amis_freebsd12_latest)
    p6_cirrus_inc_instance_create "jenkins" "$ami_id" "m4.large"
}

######################################################################
#<
#
# Function: str key_id = p6_aws_kms_svc_key_make(account_id, key_description, key_alias)
#
#  Args:
#	account_id -
#	key_description -
#	key_alias -
#
#  Returns:
#	str - key_id
#
#>
######################################################################
p6_aws_kms_svc_key_make() {
    local account_id="$1"
    local key_description="$2"
    local key_alias="$3"

    local key_admin_principals="arn:aws:iam::${account_id}:role/SSO/SSO_Admin"
    local key_user_principals="arn:aws:iam::${account_id}:role/SSO/SSO_Admin"

    local key_policy=$(p6_aws_util_template_process "iam/kms" "ACCOUNT_ID=$account_id" "KEY_ADMIN_PRINCIPALS=$key_admin_principals" "KEY_USER_PRINCIPALS=$key_user_principals")

    local key_id=$(p6_aws_cmd kms create-key "$key_description" "$key_policy")
    p6_aws_cmd kms alias-key "$key_alias" "$key_id"

    p6_return_str "$key_id"
}